### PR TITLE
Nix the global `TYPE_ENGINE`

### PR DESCRIPTION
**Pull request by Centril**
First part of https://github.com/FuelLabs/sway/issues/2063.

The goal here is to remove all the functions that use the lazy static `TYPE_ENGINE`, and once that's done, we can safely remove it.

Unfortunately, we have to introduce `PartialEqWithTypeEngine` and friends, but eventually we should be able to revert back to good ol' `PartialEq` once https://github.com/FuelLabs/sway/issues/3115 is done.

### CODE REVIEW

**Review by mohammadfawaz — DISMISSED** [review]

File: sway-parse/src/lib.rs [hunk 1]
```diff
@@ -37,3 +37,12 @@ pub fn parse_file(
     let ts = lex(handler, &src, 0, src.len(), path)?;
     Parser::new(handler, &ts).parse_to_end().map(|(m, _)| m)
 }
+
+pub fn parse_module_kind(
```
Comment (emilyaherbert) [inline sway-parse/src/lib.rs:41]: Is this a related change?

File: sway-core/src/type_system/type_engine.rs [hunk 2]
```diff
@@ -1,29 +1,48 @@
...
 use crate::{
     concurrent_slab::ConcurrentSlab, declaration_engine::*, language::ty, namespace::Path,
     type_system::*, Namespace,
 };
 
-use lazy_static::lazy_static;
 use sway_error::{error::CompileError, type_error::TypeError, warning::CompileWarning};
 use sway_types::{span::Span, Ident, Spanned};
 
-lazy_static! {
```
Comment (emilyaherbert) [inline sway-core/src/type_system/type_engine.rs:13]: Yesss

**Review by emilyaherbert — APPROVED** [review]
> Unfortunately, we have to introduce `PartialEqWithTypeEngine` and friends, but eventually we should be able to revert back to good ol' PartialEq once #3115 is done.

Yes, adding `PartialEqWithTypeEngine` and such is not ideal but is needed for this PR in the meantime.

**Review by emilyaherbert — APPROVED** [review]

**Review by mohammadfawaz — APPROVED** [review]
